listener.go

```diff
@@ -0,0 +1,409 @@
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	if auth != "" {
+		req.Header.Set("Authorization", auth)
+	}
+
+	resp, err := l.client.Do(req)
+	if err != nil {
+		l.metrics.Registration.incRequestFailure()
+		return multierr.Combine(err, fmt.Errorf("%w: unable to make a http request", ErrRegistrationFailed))
+	}
+	defer resp.Body.Close()
+
+	l.metrics.Registration.incStatusCode(resp.StatusCode)
+
+	if resp.StatusCode == http.StatusOK {
+		return nil
+	}
+
+	rBody, _ := io.ReadAll(resp.Body)
+
+	l.logger.Info("failed to register webhook",
+		zap.Int("StatusCode", resp.StatusCode),
+		zap.ByteString("body", rBody))
+
+	return fmt.Errorf("%w: http status code(%d) was not 200",
+		ErrRegistrationFailed, resp.StatusCode)
+}
+
+// Tokenize parses the token from the request header.  If the token is not found
+// or is invalid, an error is returned.
+func (l *Listener) Tokenize(r *http.Request) (*Token, error) {
+	headers := r.Header.Values(xmidtHeader)
+	if len(headers) != 0 {
+		l.metrics.TokenHeaderUsed.inc(xmidtHeader)
+	} else {
+		headers = r.Header.Values(webpaHeader)
+		l.metrics.TokenHeaderUsed.inc(webpaHeader)
+	}
+
+	if len(headers) == 0 {
+		l.metrics.TokenOutcome.incNoTokenHeader()
+	}
+
+	choices := map[string]string{
+		"none": "",
+	}
+	list := make([]string, 0, len(headers))
+	list = append(list, "none")
+
+	for _, header := range headers {
+		header = strings.TrimSpace(header)
+		if header == "" {
+			continue
+		}
+		parts := strings.Split(header, "=")
+		if len(parts) != 2 {
+			l.metrics.TokenOutcome.incInvalidHeaderFormat()
+			return nil, ErrInvalidAuth
+		}
+
+		alg := strings.ToLower(strings.TrimSpace(parts[0]))
+		val := strings.TrimSpace(parts[1])
+		if alg == "" || val == "" {
+			l.metrics.TokenOutcome.incInvalidHeaderFormat()
+			return nil, ErrInvalidAuth
+		}
+
+		choices[alg] = val
+		list = append(list, alg)
+	}
+
+	l.metrics.TokenAlgorithms.inc(list)
+
+	best, err := l.best(list)
+	if err != nil {
+		l.metrics.TokenOutcome.incAlgorithmNotFound()
+		return nil, err
+	}
+
+	l.metrics.TokenAlgorithmUsed.inc(best)
+	l.metrics.TokenOutcome.incValid()
+
+	return NewToken(best, choices[best]), nil
+}
+
+// Authorize validates that the request body matches the hash and secret provided
+// in the token.
+func (l *Listener) Authorize(r *http.Request, t Token) error {
+	secret, err := hex.DecodeString(t.Principal())
+	if err != nil {
+		l.metrics.Authorization.incInvalidSignature()
+		return fmt.Errorf("%w: unable to decode signature", ErrInput)
+	}
+
+	if r.Body == nil {
+		l.metrics.Authorization.incEmptyBody()
+		return fmt.Errorf("%w: empty request body", ErrInput)
+	}
+
+	msg, err := io.ReadAll(r.Body)
+	r.Body.Close()
+	if err != nil {
+		l.metrics.Authorization.incUnableToReadBody()
+		return fmt.Errorf("%w: unable to read request body", ErrInput)
+	}
+
+	// Reset the body so it can be read again later.
+	r.Body = io.NopCloser(bytes.NewReader(msg))
+
+	if len(msg) == 0 {
```

Comment:
how would this happen on a reset?  Don't we already check if the body is empty further up?

Reply:
r.Body is an io.ReadCloser which could be nil or it is an object.  If it is an object, it could still contain 0 bytes to be read.  Test https://github.com/xmidt-org/wrp-listener/pull/120/files/3a20e9fbdd5975479a7b87ab7070d3694e1ffa4b#diff-fd62764a6fa8144a271deeddca51c116ed2b51bb120940df885a559bef986e62R299 shows an empty body, where https://github.com/xmidt-org/wrp-listener/pull/120/files/3a20e9fbdd5975479a7b87ab7070d3694e1ffa4b#diff-fd62764a6fa8144a271deeddca51c116ed2b51bb120940df885a559bef986e62R309 (where the http.Request.Body is not set & is nil) shows a nil body.